harness/api_checkers.go

```diff
@@ -0,0 +1,493 @@
+	}
+	span := s.tracer.StartSpan("op")
+
+	// binary extract
+	ctx, err := span.Tracer().Extract(opentracing.Binary, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not io.Reader should return error")
+	s.Nil(ctx)
+
+	// text extract
+	ctx, err = span.Tracer().Extract(opentracing.TextMap, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not TextMapReader should return error")
+	s.Nil(ctx)
+
+	// HTTP extract
+	ctx, err = span.Tracer().Extract(opentracing.HTTPHeaders, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not TextMapReader should return error")
+	s.Nil(ctx)
+
+	span.Finish()
+}
+
+// TestMultiBaggage tests calls to set multiple baggage items, and if the CheckBaggageValues option
+// is set, asserts that a baggage value was successfully retrieved from the span's SpanContext.
+// It also ensures that returning false from the ForeachBaggageItem handler aborts iteration.
+func (s *APICheckSuite) TestMultiBaggage() {
+	span := s.tracer.StartSpan("op")
+	assertEmptyBaggage(s.T(), span.Context())
+
+	span.SetBaggageItem("Bag1", "BaggageVal1")
+	span.SetBaggageItem("Bag2", "BaggageVal2")
+	if s.opts.CheckBaggageValues {
+		s.Equal("BaggageVal1", span.BaggageItem("Bag1"))
+		s.Equal("BaggageVal2", span.BaggageItem("Bag2"))
+		called := false
+		span.Context().ForeachBaggageItem(func(k, v string) bool {
+			s.False(called) // should only be called once
+			called = true
+			return false
+		})
+		s.True(called)
+	} else {
+		s.T().Log("CheckBaggageValues capability not set, skipping")
+	}
+	span.Finish()
+}
+
+// TestHTTPBaggage tests whether serializing multiple baggage items that use the same key name
+// with different casing will overwrite each other (since HTTP Header names are case-insensitive).
+func (s *APICheckSuite) TestHTTPBaggage() {
+	if !s.opts.CheckBaggageValues {
+		s.T().Skip("CheckBaggageValues capability not set, skipping")
+	}
+	// set baggage items: should result in two baggage items when serializing using HTTP headers
+	span := s.tracer.StartSpan("Bender")
+	span.SetBaggageItem("BagItem", "Val1")
+	span.SetBaggageItem("bagitem", "Val2")
+	span.SetBaggageItem("BAGITEM", "Val3")
+	span.SetBaggageItem("KeyName", "Val4")
+
+	carrier := opentracing.HTTPHeadersCarrier{}
+	err := span.Tracer().Inject(span.Context(), opentracing.HTTPHeaders, carrier)
+	s.NoError(err)
+
+	extractedContext, err := s.tracer.Extract(opentracing.HTTPHeaders, carrier)
+	if s.opts.CheckExtract {
+		s.NoError(err)
+	} else {
+		s.T().Log("CheckExtract capability not set, skipping")
+	}
+	// check baggage items
+	count := 0
+	items := make(map[string]string)
+	extractedContext.ForeachBaggageItem(func(k, v string) bool {
+		count++
+		items[k] = v
+		return true
+	})
+	s.Len(items, 2, "HTTP Baggage propagation seems to depend on case-sensitive HTTP header field names")
+	s.Equal(count, 2, "HTTP Baggage propagation seems to depend on case-sensitive HTTP header field names")
```

Comment:
hm, I am not sure about this one. It is indeed how Jaeger's baggage implemented, but case-insensitivity was in the spec some time ago and has since been removed, baggage keys are now any Unicode strings.

Reply:
Oh cool, I wasn't aware of that — I was probably reading basictracer and jaeger — how does that work with HTTP headers? I could get rid of the test but I don't know that it's good practice for a tracer to try and send two different HTTP headers with different casing because field names are supposed to be case-insensitive.

Reply:
If you are asking how it works in basic or Jaeger, the answer is it doesn't. Unicode keys are nice in theory, but we don't recommend then to users. I'm not sure if Lightstep supports those. We also recommend lower-snake-case keys only.

It's not too difficult to support arbitrary keys, but they would have to be encoded, which makes them unreadable when inspecting the traffic/headers. 

For tracers that use a binary blob the arbitrary keys are easy. We choose to keep baggage in clear text and defer Unicode support until someone really complains. 

Of course if we make that as part of the harness we'd need to fix that in Jaeger :-)

Reply:
OK, I'll remove this, thanks for the info. I don't see any mention of Unicode in https://github.com/opentracing/specification/blob/master/specification.md#set-a-baggage-item but if we want to require tracers to live up to that spec we could bring back a test like this that exercises weird mixed-case and Unicode baggage key names requiring encoding for correct propagation.